File: ch05/35/FirstNotRepeatingChar.cpp
class Solution {
public:
    /*
    int solu1(string str){//第一种解法，时间复杂度为n^2
        int len = str.length();
        if(len == 0)
            return -1;
        int i = 0;
        vector<char> exist;
        for(i; i< len; i++){
            char current = str[i];
            int j;
            for(j = 0; j < len; j++){
                if(j==i)
                    continue;
                if(str[j] == current)
                    break;
            }
            if(j == len)
                break;
        }
        return i;
    }
    */
     
    int solu2(string str){//第二种解法，用哈希表，需要n的辅助空间，但是时间复杂度是n
        if(str.length() == 0)
            return -1;
        const int hashTableLen = 256;
        int hashTable[hashTableLen] = {0};
        for(int i = 0; i<str.length(); i++){
            hashTable[str[i]] ++;
        }
         
        for(int j = 0; j < str.length(); j++){
            if(hashTable[str[j]] == 1 )
                return j;
        }
        return -1;
    }
     
    int FirstNotRepeatingChar(string str) {
         
        return solu2(str);
    }
     
     
}

File: ch05/35/FirstNotRepeatingChar.py
# -*- coding:utf-8 -*-
####题目#####
# 在一个字符串(1<=字符串长度<=10000，全部由字母组成)中找到第一个只出现一次的字符,并返回它的位置
#####解法####
# 用数组实现一个简单的哈希表，键是字符对应的acssi，值是出现的次数

class Solution:
    def FirstNotRepeatingChar(self, s):
        # write code here
        re = -1
        if len(s) == 0:
            return re

        hashTable = [0 for i in range(256)]
        for ch in s:
            hashTable[ord(ch)] += 1

        for i in range(len(s)):
            # if hashTable[ord(s[i])] != 0:
            #     print hashTable[ord(s[i])]

            if hashTable[ord(s[i])] == 1:
                re = i
                break
        
        return re

if __name__ == "__main__":
    s  = "google"
    solu = Solution()
    print solu.FirstNotRepeatingChar(s)
            
        
